File: dynamic_programming/matrix_multiplication.cpp
// arr 存放 每个矩阵的维度 第 i 个矩阵的维度为 arr[i-1] * a[i]，那么给定 arr 的大小: size，则举着个数有 size-1 个
// M[i][j] 表示 A[i]A[i+1]...A[j] 的连乘最小代价
// 那么，最终要求的是 A[1]A[2]...A[n-1] 即 M[1][n-1]

#include <iostream>
#include <climits>

using namespace std;

int MatrixChainOrder(int p[], int n)
{
    int M[n][n];
    int i, j, k, L, min;

    for (i = 1; i < n; ++ i)
        M[i][i] = 0;

    // L is length
    for (L = 2; L < n; ++ L)
        for (i = 1; i+L-1 < n; ++ i)
        {
            j = i+L-1;
            M[i][j] = INT_MAX;

            for (k = i; k+1 <= j; ++ k)
            {
                min = M[i][k] + M[k+1][j] + p[i-1]*p[k]*p[j];
                if (min < M[i][j])
                    M[i][j] = min;
            }
        }

    return M[1][n-1];
}

int main()
{
    int arr[] = {1, 2, 3, 4};
    int size = sizeof(arr)/sizeof(arr[0]);
    cout << MatrixChainOrder(arr, size) << endl;

    return 0;
}

File: dynamic_programming/matrix_multiplication.py
# dp[i][j] 表示第i个矩阵与第j个矩阵所有矩阵相称所需的最少乘法次数
def matrix_chain_order(arr):
    num_matrix = len(arr) - 1;

    dp = [[0]*num_matrix for _ in range(num_matrix)]

    for n in range(2, len(arr)):
        for i in range(0, num_matrix+1-n):
            j = i + n -1
            dp[i][j] = float('inf')
            for k in range(i, j):
                min = dp[i][k] + dp[k+1][j] + arr[i] * arr[k+1] * arr[j+1]
                dp[i][j] = min if min < dp[i][j] else dp[i][j]

    return dp[0][num_matrix - 1], dp

def main():
    arr = [1, 2, 3, 4]
    min_mul, dp = matrix_chain_order(arr)
    print("Minimum number of multiplication is {}".format(min_mul))
    print(dp)

if __name__ == "__main__":
    main()
